File: packages/react-native-autoplay/ios/hybrid/HybridCluster.swift
//
//  HybridCluster.swift
//  Pods
//
//  Created by Manuel Auer on 25.10.25.
//
import NitroModules

class HybridCluster: HybridHybridClusterSpec {
    private static var listeners = [
        ClusterEventName: [String: (_:String) -> Void]
    ]()

    private static var eventQueue = [
        ClusterEventName: [String]  // clusterIds queued per event
    ]()

    private static var colorSchemeListeners = [
        String: (_: String, _: ColorScheme) -> Void
    ]()

    func addListener(
        eventType: ClusterEventName,
        callback: @escaping (_ clusterId: String) -> Void
    ) throws -> () -> Void {
        let uuid = UUID().uuidString
        HybridCluster.listeners[eventType, default: [:]][uuid] =
            callback

        if let queuedClusterIds = HybridCluster.eventQueue[eventType] {
            for clusterId in queuedClusterIds {
                callback(clusterId)
            }
            HybridCluster.eventQueue[eventType] = nil
        }

        return {
            HybridCluster.listeners[eventType]?.removeValue(
                forKey: uuid
            )
        }
    }

    func initRootView(clusterId: String) throws -> Promise<Void> {
        return Promise.async {
            if #available(iOS 15.4, *) {
                try await MainActor.run {
                    let scene = try SceneStore.getClusterScene(
                        clusterId: clusterId
                    )
                    scene?.initRootView()
                }
            } else {
                throw AutoPlayError.unsupportedVersion(
                    "Cluster support only available on iOS >= 15.4"
                )
            }
        }
    }

    func setAttributedInactiveDescriptionVariants(
        clusterId: String,
        attributedInactiveDescriptionVariants:
            [NitroAttributedString]
    ) throws {
        if #available(iOS 15.4, *) {
            let scene = try SceneStore.getClusterScene(
                clusterId: clusterId
            )
            scene?.setAttributedInactiveDescriptionVariants(
                attributedInactiveDescriptionVariants:
                    attributedInactiveDescriptionVariants
            )
        } else {
            throw AutoPlayError.unsupportedVersion(
                "Cluster support only available on iOS >= 15.4"
            )
        }
    }

    func addListenerColorScheme(
        callback: @escaping (String, ColorScheme) -> Void
    ) throws -> () -> Void {
        let uuid = UUID().uuidString
        HybridCluster.colorSchemeListeners[uuid] = callback

        return {
            HybridCluster.colorSchemeListeners.removeValue(
                forKey: uuid
            )
        }
    }

    static func emit(event: ClusterEventName, clusterId: String) {
        guard let listeners = HybridCluster.listeners[event], !listeners.isEmpty
        else {
            // no listeners -> queue the event
            HybridCluster.eventQueue[event, default: []].append(clusterId)
            return
        }

        listeners.values.forEach {
            $0(clusterId)
        }
    }
    
    static func emitColorScheme(clusterId: String, colorScheme: ColorScheme) {
        HybridCluster.colorSchemeListeners.values.forEach {
            $0(clusterId, colorScheme)
        }
    }
}

File: packages/react-native-autoplay/android/src/main/java/com/margelo/nitro/at/g4rb4g3/autoplay/hybrid/HybridCluster.kt
package com.margelo.nitro.at.g4rb4g3.autoplay.hybrid

import com.margelo.nitro.at.g4rb4g3.autoplay.AndroidAutoSession
import com.margelo.nitro.at.g4rb4g3.autoplay.VirtualRenderer
import com.margelo.nitro.at.g4rb4g3.autoplay.utils.ThreadUtil
import com.margelo.nitro.core.Promise

class HybridCluster : HybridHybridClusterSpec() {
    override fun addListener(
        eventType: ClusterEventName, callback: (String) -> Unit
    ): () -> Unit {
        val callbacks = listeners.getOrPut(eventType) { mutableListOf() }
        callbacks.add(callback)

        eventQueue[eventType]?.forEach {
            callback(it)
        }
        eventQueue[eventType] = emptyArray<String>()

        return {
            listeners[eventType]?.removeAll { it === callback }
        }
    }

    override fun initRootView(clusterId: String): Promise<Unit> {
        return Promise.async {
            if (VirtualRenderer.hasRenderer(clusterId)) {
                return@async
            }
            val carContext = AndroidAutoSession.getCarContext(clusterId)
                ?: throw IllegalArgumentException("initRootView failed, carContext for $clusterId cluster not found")
            val result = ThreadUtil.postOnUiAndAwait {
                VirtualRenderer(carContext, clusterId)
            }
            if (result.isFailure) {
                throw result.exceptionOrNull()
                    ?: UnknownError("unknown error initializing the virtual screen")
            }
        }
    }

    override fun setAttributedInactiveDescriptionVariants(
        clusterId: String, attributedInactiveDescriptionVariants: Array<NitroAttributedString>
    ) {
        throw IllegalAccessException("setAttributedInactiveDescriptionVariants is supported on iOS only")
    }

    override fun addListenerColorScheme(callback: (String, ColorScheme) -> Unit): () -> Unit {
        colorSchemeListeners.add(callback)

        return {
            colorSchemeListeners.remove(callback)
        }
    }

    override fun addListenerZoom(callback: (String, ZoomEvent) -> Unit): () -> Unit {
        throw IllegalAccessException("addListenerZoom is supported on iOS only")
    }

    override fun addListenerCompass(callback: (String, Boolean) -> Unit): () -> Unit {
        throw IllegalAccessException("addListenerCompass is supported on iOS only")
    }

    override fun addListenerSpeedLimit(callback: (String, Boolean) -> Unit): () -> Unit {
        throw IllegalAccessException("addListenerSpeedLimit is supported on iOS only")
    }

    companion object {
        const val TAG = "HybridCluster"

        private val listeners =
            mutableMapOf<ClusterEventName, MutableList<(clusterId: String) -> Unit>>()
        private val eventQueue = mutableMapOf<ClusterEventName, Array<String>>()

        private val colorSchemeListeners =
            mutableListOf<(clusterId: String, colorScheme: ColorScheme) -> Unit>()

        fun emit(event: ClusterEventName, clusterId: String) {
            if (listeners[event].isNullOrEmpty()) {
                val clusterIds = eventQueue.getOrPut(event) { emptyArray<String>() }
                eventQueue[event] = clusterIds.plus(clusterId)
                return
            }

            listeners[event]?.forEach { it(clusterId) }
        }

        fun emitColorScheme(clusterId: String, colorScheme: ColorScheme) {
            colorSchemeListeners.forEach {
                it(clusterId, colorScheme)
            }
        }
    }
}